### PR TITLE
🚨🚨 Source Mailchimp: Migrate to Low code

### PR DESCRIPTION
**Pull request by artem1205**
## What

Resolve https://github.com/airbytehq/airbyte-internal-issues/issues/2919

## How

migrate to Low-Code

## Recommended reading order
1. `airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml`
2. `airbyte-integrations/connectors/source-mailchimp/source_mailchimp/components.py`

## 🚨 User Impact 🚨

- breaking change for nested state (parent child incremental)
- fix segment_members primary key: ["id", "segment_id"] 
- fix list_members primary key: ["id", "list_id"] 

## Pre-merge Actions

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

### CODE REVIEW

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/components.py
Comment (natikgadzhi) [inline airbyte-integrations/connectors/source-mailchimp/source_mailchimp/components.py:117]: Question so I learn: looks like this filter takes a dict of record and returns a separate row for each record and each activity in that record. 

If there are no activities at all, the record will be filtered out. If there are multiple activities on a single record (if that is possible), that will return multiple row for each found activity item. 

How wrong am I?

If that's correct, than this thing both _filters_ records, but also _transforms_ them and generates them, right?

Comment (girarda) [reply airbyte-integrations/connectors/source-mailchimp/source_mailchimp/components.py:120]: why is this defined as being a filter? Like Natik mentioned, I would expect this to be a custom transformation followed by a filter that checks if there were activities.

Separating the two concerns allows us to extract some logic that can be generalized.

Instead of creating a custom filter, we should introduce a flatten_field transformation which takes a nested object and brings it to the root

The transformation:
```
{
"key": 123,
"activities": {"nested": "value", "another_nested": "value2"} 
}

to 
{
"key": 123,
"nested:" value",
"another_nested": "value2"
}
```

would then be described as
```
transformations:
  type: FlattenField
  field_pointers:
    - activities
```

Here is a tentative schema definition:
```
  FlattenFields:
    title: Flatten Fields
    type: object
    required:
      - type
      - field_pointers
    properties:
      type:
        type: string
        enum: [FlattenFields]
      field_pointers:
        title: Field Paths
        type: array
        items:
          items:
            type: string
        examples:
          - ["activities"]
```

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/components.py
Comment (girarda) [inline airbyte-integrations/connectors/source-mailchimp/source_mailchimp/components.py:20]: can you describe why this is needed? it's difficult to judge whether is should truly be custom behavior without stating the desired behavior and why it cannot be supported natively by the CDK

File: airbyte-integrations/connectors/source-mailchimp/unit_tests/test_source.py
Comment (girarda) [inline airbyte-integrations/connectors/source-mailchimp/unit_tests/test_source.py:47]: have we considered the tradeoffs between deleting the tests and updating the fixture to run them in the new implementation?

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/components.py
Comment (artem1205) [reply airbyte-integrations/connectors/source-mailchimp/source_mailchimp/components.py:120]: Hmm, I can agree that this is more about `transformation` rather than `filtering`, but 

```json
{
"key": 123,
"activities": [
{"nested": "value"},
 {"another_nested": "value2"}]
}

```
should be transformed to 2 records as `activities` is a list :
```json
[
  {
    "key": 123,
    "nested": "value"
  },
  {
    "key": 123,
    "another_nested": "value2"
  }
]
```
it means that we will have more records than original list, while current implementation does not expect the number of records to change:

https://github.com/airbytehq/airbyte/blob/4061f08f3d34e2de651c8d2a445952f3335ebcc9/airbyte-cdk/python/airbyte_cdk/sources/declarative/extractors/record_selector.py#L88-L97

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/components.py
Comment (girarda) [inline airbyte-integrations/connectors/source-mailchimp/source_mailchimp/components.py:52]: why are we modifying the config? As far as I know, this is an implementation detail. Why not remove the url_base field from the manifest?

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/components.py
Comment (girarda) [reply airbyte-integrations/connectors/source-mailchimp/source_mailchimp/components.py:120]: got it! I did misunderstand. 

Can you make a proposal for how you'd want to use this feature? When considering new features or custom components, it's useful to have a description, examples, and a desired interface (when known) to help us understand the use case. 

Then we can implement and make any required changes to the record transform interface

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/components.py
Comment (artem1205) [reply airbyte-integrations/connectors/source-mailchimp/source_mailchimp/components.py:77]: agreed, fixed

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/components.py
Comment (artem1205) [reply airbyte-integrations/connectors/source-mailchimp/source_mailchimp/components.py:20]: So, mailchimp declares custom [api structure](https://mailchimp.com/developer/marketing/docs/fundamentals/#api-structure)

> There are a few ways to find your data center. It’s the first part of the URL you see in the[ API keys section](https://us1.admin.mailchimp.com/account/api/) of your account; if the URL is https://us6.mailchimp.com/account/api/, then the data center subdomain is us6. It’s also appended to your API key in the form key-dc; if your API key is `0123456789abcdef0123456789abcde-us6`, then the data center subdomain is us6. And finally, if you’re connecting via OAuth 2, you can find the data center associated with the token via the OAuth Metadata endpoint; for more information, see the [OAuth guide](https://mailchimp.com/developer/marketing/guides/access-user-data-oauth-2/#implement-the-oauth-2-workflow-on-your-server).

The only reason for this custom component is to get the `data center` prefix for Oauth authenticator.

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/components.py
Comment (artem1205) [reply airbyte-integrations/connectors/source-mailchimp/source_mailchimp/components.py:52]: Mainly for `visibility` and not overriding the `get_url_base` method from base class

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/components.py
Comment (artem1205) [reply airbyte-integrations/connectors/source-mailchimp/source_mailchimp/components.py:107]: Sure, I'll get back with all propositions as soon as this PR will be in review, I'm still working on it.

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml
Comment (maxi297) [inline airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml:279]: Differences observed:
* Empty strings were removed from from `merge_fields`
* The parent stream `lists_stream` was not sliced in the previous implementation
* `segments` was not sliced in the previous implementation

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml
Comment (maxi297) [inline airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml:315]: Differences observed:
* The new implementation does not consider `config["campaign_id"]` hence giving 462 records instead of 2

File: airbyte-integrations/connectors/source-mailchimp/acceptance-test-config.yml
Comment (maxi297) [inline airbyte-integrations/connectors/source-mailchimp/acceptance-test-config.yml:15]: Why remove this?

File: airbyte-integrations/connectors/source-mailchimp/acceptance-test-config.yml
Comment (maxi297) [inline airbyte-integrations/connectors/source-mailchimp/acceptance-test-config.yml:36]: Why remove this?

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/components.py
Comment (maxi297) [reply airbyte-integrations/connectors/source-mailchimp/source_mailchimp/components.py:22]: Why is this a custom requester instead of something like a config migrator?

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/components.py
Comment (maxi297) [reply airbyte-integrations/connectors/source-mailchimp/source_mailchimp/components.py:54]: I would agree with @girarda that it would be better to pass this as a parameter of `__init__` else we create an object that is stateful. That being said, I don't know how all of this fits in the initialization process of the custom components. The other solution I would propose is to not use custom components and use only interpolation

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/components.py
Comment (maxi297) [inline airbyte-integrations/connectors/source-mailchimp/source_mailchimp/components.py:93]: Note to @girarda: semi-incremental needs to work with per partition states

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/components.py
Comment (maxi297) [inline airbyte-integrations/connectors/source-mailchimp/source_mailchimp/components.py:94]: Why is this TODO needed? Shouldn't we just accept everything if there are no states (which is the logic you implemented here)?

**Review by maxi297 — CHANGES_REQUESTED** [review]
Can you address the comments and I'll have another review after that. Thanks!

File: airbyte-integrations/connectors/source-mailchimp/acceptance-test-config.yml
Comment (artem1205) [reply airbyte-integrations/connectors/source-mailchimp/acceptance-test-config.yml:36]: both configs (oauth and apikey) works with the same recordset. We already have a test_connection to ensure them.
In my opinion second test_read is excessive.

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/components.py
Comment (artem1205) [reply airbyte-integrations/connectors/source-mailchimp/source_mailchimp/components.py:94]: agreed

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml
Comment (artem1205) [reply airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml:6]: i've tried to, but unfortunately, streams to check should be declared in `streams` list

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml
Comment (artem1205) [reply airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml:36]: thanks for notice, my bad, will change

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml
Comment (artem1205) [reply airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml:91]: Hmm, interesting, I'll make a P100Y step as it is done in other incremental settings.

**Comment by artem1205** [comment]
UPD: Requester,
I've remembered why did I implemented in this way.
As far as custom components are created for each stream separately, we'll make extra requests to mailchimp just to get the datacenter (1 request per stream). If we'll update the configuration (like it is now) only 1 request will be made.

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml
Comment (artem1205) [reply airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml:110]: > it feels like a breaking change if empty strings are now

it is not a breaking change, since no schema differences introduced here. I thought that RemoveField is the closest type of transformation that can be applied here.

Comment (maxi297) [reply airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml:110]: It is not in respect to the schema but it might be in records to the user's consumption. For example, if the user does `select X from Y where address == ""`, his workflow won't work anymore

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml
Comment (maxi297) [reply airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml:91]: Actually, if `step` is not provided, we use [datetime.timedelta.max](https://github.com/airbytehq/airbyte/blob/27d9207ce4e662deef9438b225629456368fca79/airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/datetime_based_cursor.py#L79-L83) so we can simply remove the step in that case

File: docs/integrations/sources/mailchimp-migrations.md
Comment (maxi297) [inline docs/integrations/sources/mailchimp-migrations.md:5]: Can we be explicit as to why we need a reset?

**Comment by artem1205** [comment]
@maxi297 ,

> Why is this a custom requester instead of something like a config migrator?

Maybe I miss something, could you provide more details on how config migration will help here?

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml
Comment (artem1205) [reply airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml:119]: fixed, got 51 records in latest sync

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml
Comment (artem1205) [reply airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml]: agreed, we can have the same member under the different segments.
changed primary key to composite one

File: docs/integrations/sources/mailchimp-migrations.md
Comment (artem1205) [reply docs/integrations/sources/mailchimp-migrations.md:7]: updated

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml
Comment (artem1205) [reply airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml:136]: updated, slicing removed

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml
Comment (artem1205) [reply airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml:168]: updated, slicing removed

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml
Comment (artem1205) [reply airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml:146]: updated, slicing removed

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml
Comment (artem1205) [reply airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml:190]: updated, slicing removed

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml
Comment (artem1205) [reply airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml:217]: updated, slicing removed

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml
Comment (artem1205) [reply airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml:275]: updated, slicing removed

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml
Comment (artem1205) [reply airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml:309]: campaign id was added as a `hidden` field to config and used in CAT tests only.
I think that CAT are already fast enough not to add this limitation.
Another cons is the structure of YAML:
we need to add another custom component that will choose `type` of `partition_router` based on config["campaign_id"] presence: `SubstreamPartitionRouter` or `ListPartitionRouter`.

I'd prefer not to add this custom component just to cover the test.

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml
Comment (maxi297) [reply airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml:6]: It seems like [this line](https://github.com/airbytehq/airbyte/blob/f67938993e687a4b2ce2fed9ff81315f7f92e2cd/airbyte-cdk/python/airbyte_cdk/sources/declarative/checks/check_stream.py#L31) assumes the stream we test is indeed part of the catalog. 

For this specific case, are we losing information using a stream instead of a ping? I think we have some sources that queries an endpoint to retrieve the scopes of the auth token and prints the missing scopes for example which is clearly added value. What is the added value we would be losing in this case?

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml
Comment (maxi297) [reply airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml:36]: I don't see the change yet. Am I missing something?

Comment (artem1205) [reply airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml:36]: i've changed line 40:
field_name: "count"

File: airbyte-integrations/connectors/source-mailchimp/acceptance-test-config.yml
Comment (artem1205) [reply airbyte-integrations/connectors/source-mailchimp/acceptance-test-config.yml:15]: it is not relevant, and causes error in selectiveauthenticator

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml
Comment (artem1205) [reply airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml:6]: Not in this case, I'll remove todo to avoid confusion

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml
Comment (maxi297) [reply airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml:36]: My bad, I was checking the line on the comment and not the one on the RequestOption. Thanks for pointing that out!

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml
Comment (maxi297) [reply airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml:91]: It seems like we still don't have a lookback window. We will face the same issue as the one described [here](https://airbytehq-team.slack.com/archives/C02U9R3AF37/p1705335177760819).

**Comment by maxi297** [comment]
> UPD: Requester, I've remembered why did I implemented in this way. As far as custom components are created for each stream separately, we'll make extra requests to mailchimp just to get the datacenter (1 request per stream). If we'll update the configuration (like it is now) only 1 request will be made.

Can't this just be config migration that happens at the beginning of every sync?

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml
Comment (maxi297) [reply airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml:309]: I didn't know the field was for CATs only. Makes sense to remove it than, thanks!

**Review by maxi297 — COMMENTED** [review]
Last thing I would like is to remove the custom requester and use a config migration at the beginning of the sync. The reason I prefer this is that it reduce the number of custom components which improves maintainability. As the goal of moving to low-code is improve maintainability, I fear this might be going in the wrong direction. Would that makes sense?

**Comment by artem1205** [comment]
> Can't this just be config migration that happens at the beginning of every sync?

I think so, I can move this functionality to config migration and update this field every time we run the sync.

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml
Comment (artem1205) [reply airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml:87]: added lookback

**Comment by artem1205** [comment]
UPD: Requester,
removed custom component, moved logic to config_migration:
script will add data_center property on every start

File: airbyte-integrations/connectors/source-mailchimp/source_mailchimp/config_migrations.py
Comment (maxi297) [inline airbyte-integrations/connectors/source-mailchimp/source_mailchimp/config_migrations.py:81]: This whole method can be replaced by `print(create_connector_config_control_message(migrated_config).json(exclude_unset=True))` and we can remove the message_repository from this class

**Review by maxi297 — APPROVED** [review]
Thanks for addressing all the comments and making sure there is a little impact as possible for the user!

**Review by lazebnyi — APPROVED** [review]

**Comment by NAjustin** [comment]
@artem1205 I found this PR while investigating an issue with the current version of the Mailchimp connector, and noticed that this version replicates the same error, so I wanted to flag it for you while I work on testing a PR for the main connector.

**The issue is that the `members` object currently has `primary_key` set to `id`, which is incorrect:**
- In Mailchimp, `members` are associated with `lists`. This means there will be one member record per list they are on.
- `members`.id` is a hash of the member's email address, which is obviously non-unique across lists
- This means that if you do an incremental sync with the current connector (or this low-code version), you will only get one record per email.

You can see this issue [here](https://github.com/airbytehq/airbyte/blob/6e92e8ee1917f7a01510afc0530af883b0ec3388/airbyte-integrations/connectors/source-mailchimp/source_mailchimp/manifest.yaml#L129C7-L129C18) in your feature branch.

**Either of the following keys would be suitable:**
- `contact_id`, which is a non-email-specific identifier for the member within the list (so is unique across email+list)
- `web_id` is an integer and is also unique to the list+email combination
- `id` and `list_id` could also be used as a compound key with the same effect

**Mailchimp also discourages the use of `id` in their [docs](https://mailchimp.com/developer/marketing/api/list-members/list-members-info/) (see the `members`.`contact_id` response object), stating:**
>As Mailchimp evolves beyond email, you may eventually have contacts without email addresses. While the `id` is the MD5 hash of their email address, this `contact_id` is agnostic of contact’s inclusion of an email address.

This has been a longstanding issue in the Mailchimp source, but only affects accounts with the same contacts on multiple lists which may be why it's flown under the radar.

**Comment by artem1205** [comment]
@NAjustin, thank you for pointing that out!
composite primary key will be used: ["id", "list_id"]

**Review by brianjlai — COMMENTED** [review]
small reminder to re-run `poetry lock` before merging to use the latest version of `airbyte-cdk`. Thanks!

**Pull request merged**
